lib/mayaUsd/fileio/utils/meshWriteUtils.cpp

```diff
@@ -377,4 +405,76 @@ UsdMayaMeshUtil::exportReferenceMesh(UsdGeomMesh& primSchema, MObject obj)
     primVar.GetAttr().Set(VtValue(points));
 }
 
+void
+UsdMayaMeshUtil::assignSubDivTagsToUSDPrim(MFnMesh& meshFn,
+                                           UsdGeomMesh& primSchema,
+                                           UsdUtilsSparseValueWriter& valueWriter)
+{
+    // Vert Creasing
+    MUintArray mayaCreaseVertIds;
+    MDoubleArray mayaCreaseVertValues;
+    meshFn.getCreaseVertices(mayaCreaseVertIds, mayaCreaseVertValues);
+    if (!TF_VERIFY(mayaCreaseVertIds.length() == mayaCreaseVertValues.length())) {
+        return;
+    }
+    if (mayaCreaseVertIds.length() > 0u) {
+        VtIntArray subdCornerIndices(mayaCreaseVertIds.length());
+        VtFloatArray subdCornerSharpnesses(mayaCreaseVertIds.length());
+        for (unsigned int i = 0u; i < mayaCreaseVertIds.length(); ++i) {
+            subdCornerIndices[i] = mayaCreaseVertIds[i];
+            subdCornerSharpnesses[i] = mayaCreaseVertValues[i];
+        }
+
+        // not animatable
+        UsdMayaWriteUtil::SetAttribute(primSchema.GetCornerIndicesAttr(), &subdCornerIndices, valueWriter);
+
+        // not animatable
+        UsdMayaWriteUtil::SetAttribute(primSchema.GetCornerSharpnessesAttr(), &subdCornerSharpnesses, valueWriter);
+    }
+
+    // Edge Creasing
+    int edgeVerts[2];
+    MUintArray mayaCreaseEdgeIds;
+    MDoubleArray mayaCreaseEdgeValues;
+    meshFn.getCreaseEdges(mayaCreaseEdgeIds, mayaCreaseEdgeValues);
+    if (!TF_VERIFY(mayaCreaseEdgeIds.length() == mayaCreaseEdgeValues.length())) {
+        return;
+    }
+    if (mayaCreaseEdgeIds.length() > 0u) {
+        std::vector<int> subdCreaseIndices(mayaCreaseEdgeIds.length() * 2);
+        std::vector<float> subdCreaseSharpnesses(mayaCreaseEdgeIds.length());
+        for (unsigned int i = 0u; i < mayaCreaseEdgeIds.length(); ++i) {
+            meshFn.getEdgeVertices(mayaCreaseEdgeIds[i], edgeVerts);
+            subdCreaseIndices[i * 2] = edgeVerts[0];
+            subdCreaseIndices[i * 2 + 1] = edgeVerts[1];
+            subdCreaseSharpnesses[i] = mayaCreaseEdgeValues[i];
+        }
```

Comment:
It might be slightly nicer to do:
```c++
        std::vector<int> subdCreaseIndices(mayaCreaseEdgeIds.length() * 2);
        // just construct directly from the array data
        // by moving this out of the loop, you'll leverage simd ops here.
        std::vector<float> subdCreaseSharpnesses(&mayaCreaseEdgeIds[0], &mayaCreaseEdgeIds[0] + mayaCreaseEdgeIds.length());
        // avoid dso call by taking a copy of length. 
        for (unsigned int i = 0u, n = mayaCreaseEdgeIds.length(); i < n; ++i) {
            meshFn.getEdgeVertices(mayaCreaseEdgeIds[i], edgeVerts);
            subdCreaseIndices[i * 2] = edgeVerts[0];
            subdCreaseIndices[i * 2 + 1] = edgeVerts[1];
        }
```

Reply:
@robthebloke Thanks. See 8e3a0accd75df424493e2911dcb5f0a10d629092

Just so I understand, bypassing the call per iteration is mainly because the compiler has no visibility into Maya's array structure so it cannot optimize it as well as your loop? Is that what you mean by `avoid dso call`?